### DISCUSSION TITLE
Should the Scheduling Request (SR) period be a multiple of the TDD pattern period in 5G?

### DISCUSSION THREAD

**Discussion by arman-maghsoudnia** [srsran/srsRAN_Project#600] [General Help] [answered]
Dear all,

In the srsRAN implementation, the sr period must be a multiple of the TDD pattern period. My question is whether the spec enforces this limitation or is it just an srsRAN limitation. I thoroughly checked the spec to find the answer to this question, but could not find the answer. I really would appreciate it if anybody can help me with this question. 

Best,
Arman

**Comment by herlesupreeth** [comment]
If I recall correctly, its a srsRAN introduced constraint

**Reply by arman-maghsoudnia** [reply]
Thanks for the answer. Do you what spec says about this? Because I feel that the period cannot be arbitrary. Because if you set it to an arbitrary value, some occasions of SR could happen in DL slot which doesn't make sense.

**Reply by herlesupreeth** [reply]
![image](https://github.com/srsran/srsRAN_Project/assets/4115839/7b0157c9-0e63-459e-81fe-700ce1c546da)

RRC spec (TS 38.331)

> Because if you set it to an arbitrary value, some occasions of SR could happen in DL slot which doesn't make sense.

When you set the SR period in configuration, upon gNB initialization we generate SR offset based on where the UL slots are present so we ensure SR does not fall in DL slot

**Reply by arman-maghsoudnia** [reply]
Even when we set the Offset, SR falling in the DL can happen. Let me give an example:
Assume TDD pattern is DDMUU and we set the SR period to be 3. 
The SR occasion against the slots are like this (N mean no SR and S means SR):

![image](https://github.com/srsran/srsRAN_Project/assets/93431904/419d2569-9553-4812-8e0e-2c5dbc39725a)

As can be seen, some SR occasions happen in DL.

**Reply by herlesupreeth** [reply]
SR period 3msec/3slots is invalid. Also, DDMUU TDD pattern is not possible when configured via TDD-UL-DL-ConfigCommon as it equates to 1.5msec, which is not a valid value as per RRC spec

![image](https://github.com/srsran/srsRAN_Project/assets/4115839/a5a6a561-f6f9-44cd-a068-d38bbb96aca5)

**Reply by arman-maghsoudnia** [reply]
If we use 30KHz subcarrier spacing, hence 0.5ms slot duration, the DDMUU is valid because it results in 2.5ms TDD periodicity which is allowed.

```
 TDD-UL-DL-Pattern ::= SEQUENCE {
    dl-UL-TransmissionPeriodicity              ENUMERATED {ms0p5, ms0p625, ms1, 
                                                           ms1p25, ms2, ms2p5, ms5, ms10},
```

You are right about the SR period. It is not valid. But let me change it so to be valid. Now I set it to 4 slots which is valid for 30KHz subcarrier spacing as per the spec:

```
periodicityAndOffset : SR periodicity and offset in number of symbols or slots (see TS 38.213 [13], clause 9.2.4) The following periodicities may be configured depending on the chosen subcarrier spacing:

    SCS = 15 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 5sl, 8sl, 10sl, 16sl, 20sl, 40sl, 80sl

    SCS = 30 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 8sl, 10sl, 16sl, 20sl, 40sl, 80sl, 160sl

    SCS = 60 kHz: 2sym, 7sym/6sym, 1sl, 2sl, 4sl, 8sl, 16sl, 20sl, 40sl, 80sl, 160sl, 320sl

    SCS = 120 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 8sl, 16sl, 40sl, 80sl, 160sl, 320sl, 640sl

sym6or7 corresponds to 6 symbols if extended cyclic prefix and a SCS of 60 kHz are configured, otherwise it corresponds to 7 symbols.

For periodicities 2sym, 7sym and sl1 the UE assumes an offset of 0 slots.
```

Now let's take a look at the SR occasions.:

![image](https://github.com/srsran/srsRAN_Project/assets/93431904/b98ffea9-ae86-41a4-afc1-d6c0a0d53aee)

Again, some of the SR occasions fall into the DL.

**Reply by herlesupreeth** [reply]
> If we use 30KHz subcarrier spacing, hence 0.5ms slot duration, the DDMUU is valid because it results in 2.5ms TDD periodicity which is allowed.

true, my bad.

> Again, some of the SR occasions fall into the DL.

I finally recalled that this was the reason we put in the validator to ensure SR period is a multiple of TDD pattern :) Thanks for detailed example and helping me recall.

**Reply by arman-maghsoudnia** [reply]
No problem at all. But I was wondering how does the spec handle this challenge. Because after hours of reading the spec, I guess in the spec they do not put this restriction. Do you have any idea about this?

**Reply by herlesupreeth** [reply]
I think they have put that restriction in a very vague statement (below) in TS 38.213, clause 9.2.4. And, in section 11.1 it deals with TDD pattern etc.

> SR transmission occasions in a PUCCH are subject to the limitations for UE transmissions described in Subclause 11.1
> and Subclause 11.1.1.

**Reply by arman-maghsoudnia** [reply]
This is the exact path that I also went down into it to see whether they mentioned what to do or not. But I failed to find what I was looking for. So if anybody could get the answer to the question, I would appreciate it.

**Reply by herlesupreeth** [reply, marked as answer]
Technically, spec doesn't restrict what SR period to be use. But, I found that if SR scheduled over DL slot, then PUCCH wont be sent.

> A UE considers symbols in a slot indicated as downlink by tdd-UL-DL-ConfigurationCommon, or tdd-UL-DL-
> ConfigurationDedicated to be available for receptions and considers symbols in a slot indicated as uplink by tdd-UL-
> DL-ConfigurationCommon, or by tdd-UL-DL-ConfigurationDedicated to be available for transmissions.

and

> For a set of symbols of a slot that are indicated to a UE as downlink by tdd-UL-DL-ConfigurationCommon, or tdd-UL-
> DL-ConfigurationDedicated, the UE does not transmit PUSCH, PUCCH, PRACH, or SRS in the set of symbols of the
> slot.

**Reply by arman-maghsoudnia** [reply]
I see, great! Thanks for the help. My brain was so biased looking for the restriction on SR period that I missed this part. 
So we can conclude that there are no restrictions on the SR period (if it is in the allowed list for the subcarrier spacing).

```
periodicityAndOffset : SR periodicity and offset in number of symbols or slots (see TS 38.213 [13], clause 9.2.4) The following periodicities may be configured depending on the chosen subcarrier spacing:

    SCS = 15 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 5sl, 8sl, 10sl, 16sl, 20sl, 40sl, 80sl

    SCS = 30 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 8sl, 10sl, 16sl, 20sl, 40sl, 80sl, 160sl

    SCS = 60 kHz: 2sym, 7sym/6sym, 1sl, 2sl, 4sl, 8sl, 16sl, 20sl, 40sl, 80sl, 160sl, 320sl

    SCS = 120 kHz: 2sym, 7sym, 1sl, 2sl, 4sl, 8sl, 16sl, 40sl, 80sl, 160sl, 320sl, 640sl

sym6or7 corresponds to 6 symbols if extended cyclic prefix and a SCS of 60 kHz are configured, otherwise it corresponds to 7 symbols.

For periodicities 2sym, 7sym and sl1 the UE assumes an offset of 0 slots.
```

But in the case that due to the periodicity, one occasion that requires uplink transmission happens in the DL slot, the UE has to cancel that occasion. 

Thanks for the help @herlesupreeth

**Comment by frankist** [comment]
@arman-maghsoudnia, do you have any particular use-case for a SR period different than the TDD period?

**Reply by arman-maghsoudnia** [reply]
It can reduce the latency. For example, assume that the TDD pattern is DMUUUUUUU. If you set the SR period to 3 slots here, the occasions can be like this: DMSUUSUUS, where S is an SR in the UL. So the opportunities for sending SR are higher and as a result, it can reduce the latency.

But I feel it can only be beneficial when the number of UL slots is higher than the DL slots. And I think that you do not support more UL than DL either.

**Reply by herlesupreeth** [reply]
> But I feel it can only be beneficial when the number of UL slots is higher than the DL slots. And I think that you do not support more UL than DL either.

Its supported. You can try TDD pattern DDMUUUUUUU. Also, set 

```
  pusch:
    min_k2: 2
  pucch:
    min_k1: 2
  tdd_ul_dl_cfg:
    dl_ul_tx_period: 10
    nof_dl_slots: 2
    nof_dl_symbols: 10
    nof_ul_slots: 7
    nof_ul_symbols: 0
```

**Reply by arman-maghsoudnia** [reply]
I was referring to this discussion: 
https://github.com/srsran/srsRAN_Project/discussions/252#discussioncomment-7300356
So in the new update the DCI 2_0 and as a result more UL than DL is supported?

**Reply by herlesupreeth** [reply]
> So in the new update the DCI 2_0 and as a result more UL than DL is supported?

DCI 2_0 is not supported but you can change the TDD pattern as I mentioned above which gives you more UL slots than DL

**Reply by arman-maghsoudnia** [reply]
But if they cannot be scheduled, what would be the benefit of doing so? Please refer to this answer: https://github.com/srsran/srsRAN_Project/discussions/252#discussioncomment-7276255

**Reply by herlesupreeth** [reply]
> Just increasing the nof. UL slots will not do any good since DL slots are needed (sending PDCCH for UL DATA) to schedule UL data.
> 
> For example: Imagine 1 UE testing scenario and nof. DL slots is 3 and UL slots is 6. In this case the UE can be scheduled to transmit UL data in only 3 slots since the PDCCH to schedule UL data can be sent only over 3 slots so the remaining 3 slots are redundant.
> 
> So in my opinion there should be equal number of DL and UL slots.

Sorry, my understanding was wrong back then. It is possible to send multiple UL PDCCHs (over different CCEs for same UE) for scheduling PUSCHs over multiple slots. So, even if there is one DL slot we can schedule all UL slots in a TDD period. Hence, that part has been fixed in the latest release. So, whenever nof. UL slots > nof. DL slots in the TDD pattern configured, we are now using more than one k2 value to schedule all UL slots.
